README.md
# NAME

MooseX::SimpleConfig - A Moose role for setting attributes from a simple configfile

# VERSION

version 0.10

# SYNOPSIS

    ## A YAML configfile named /etc/my_app.yaml:
    foo: bar
    baz: 123

    ## In your class
    package My::App;
    use Moose;

    with 'MooseX::SimpleConfig';

    has 'foo' => (is => 'ro', isa => 'Str', required => 1);
    has 'baz'  => (is => 'rw', isa => 'Int', required => 1);

    # ... rest of the class here

    ## in your script
    #!/usr/bin/perl

    use My::App;

    my $app = My::App->new_with_config(configfile => '/etc/my_app.yaml');
    # ... rest of the script here

    ####################
    ###### combined with MooseX::Getopt:

    ## In your class
    package My::App;
    use Moose;

    with 'MooseX::SimpleConfig';
    with 'MooseX::Getopt';

    has 'foo' => (is => 'ro', isa => 'Str', required => 1);
    has 'baz'  => (is => 'rw', isa => 'Int', required => 1);

    # ... rest of the class here

    ## in your script
    #!/usr/bin/perl

    use My::App;

    my $app = My::App->new_with_options();
    # ... rest of the script here

    ## on the command line
    % perl my_app_script.pl -configfile /etc/my_app.yaml -otherthing 123

# DESCRIPTION

This role loads simple configfiles to set object attributes.  It
is based on the abstract role [MooseX::ConfigFromFile](http://search.cpan.org/perldoc?MooseX::ConfigFromFile), and uses
[Config::Any](http://search.cpan.org/perldoc?Config::Any) to load your configfile.  [Config::Any](http://search.cpan.org/perldoc?Config::Any) will in
turn support any of a variety of different config formats, detected
by the file extension.  See [Config::Any](http://search.cpan.org/perldoc?Config::Any) for more details about
supported formats.

To pass additional arguments to [Config::Any](http://search.cpan.org/perldoc?Config::Any) you must provide a
`config_any_args()` method, for example:

    sub config_any_args {
      return {
        driver_args => { General => { '-InterPolateVars' => 1 } }
      };
    }

Like all [MooseX::ConfigFromFile](http://search.cpan.org/perldoc?MooseX::ConfigFromFile) -derived configfile loaders, this
module is automatically supported by the [MooseX::Getopt](http://search.cpan.org/perldoc?MooseX::Getopt) role as
well, which allows specifying `-configfile` on the command line.

# ATTRIBUTES

## configfile

Provided by the base role [MooseX::ConfigFromFile](http://search.cpan.org/perldoc?MooseX::ConfigFromFile).  You can
provide a default configfile pathname like so:

    has '+configfile' => ( default => '/etc/myapp.yaml' );

You can pass an array of filenames if you want, but as usual the array
has to be wrapped in a sub ref.

    has '+configfile' => ( default => sub { [ '/etc/myapp.yaml', '/etc/myapp_local.yml' ] } );

Config files are trivially merged at the top level, with the right-hand files taking precedence.

# CLASS METHODS

## new\_with\_config

Provided by the base role [MooseX::ConfigFromFile](http://search.cpan.org/perldoc?MooseX::ConfigFromFile).  Acts just like
regular `new()`, but also accepts an argument `configfile` to specify
the configfile from which to load other attributes.  Explicit arguments
to `new_with_config` will override anything loaded from the configfile.

## get\_config\_from\_file

Called internally by either `new_with_config` or [MooseX::Getopt](http://search.cpan.org/perldoc?MooseX::Getopt)'s
`new_with_options`.  Invokes [Config::Any](http://search.cpan.org/perldoc?Config::Any) to parse `configfile`.

# AUTHOR

Brandon L. Black <blblack@gmail.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2007 by Brandon L. Black <blblack@gmail.com>.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

# CONTRIBUTORS

- Alex Howarth <alex.howarth@gmail.com>
- Alexander Hartmaier <alex.hartmaier@gmail.com>
- Brandon L Black <blblack@gmail.com>
- Karen Etheridge <ether@cpan.org>
- Tomas Doran <bobtfish@bobtfish.net>
- Yuval Kogman <nothingmuch@woobling.org>
- Zbigniew Lukasiak <zby@cpan.org>
- lestrrat <lestrrat+github@gmail.com>


Changes
Revision history for MooseX-SimpleConfig

{{$NEXT}}

0.11      2014-11-16 22:03:33Z
  - remove optional test prereq by testing .pl config files, rather than .conf

0.10      2013-11-13 01:13:05Z
  - repository migrated from shadowcat to the github moose organization
  - eliminate race conditions between tests by using separate tempdirs for
    config files
  - resolve failing test report by removing unnecessary and undeclared prereq

0.09      2010-08-15
  - Fix issue with MANIFEST

0.08      2010-08-13
  - Fix passing multiple config files to SimpleConfig.
    - Tests for this (zby)

0.07      2010-05-22
  - Reverse order of testing config files so that latter options on the
    command line (when multiple are specified) take precedence. (RT#57027)
  - Reduce exception thrown by being unable to find a config file to a
    warning (RT#57027)
  - Fixed configfile attribute examples
  - Added metadata to Makefile.PL

0.06      2010-04-12
  - Depend on at least version 0.13 of Config::Any for flatten_to_hash
    support.

0.05      2010-10-22
   - Allow multiple config files to be loaded and merged.
     Patch in RT#48552 from Sawyer X

0.04      2009-11-04
    - Allow options to be passed through to Config::Any more easily.

0.03      2008-01-23
    - Mostly just added pod and tests for configfile
      defaulting

0.02      2007-12-19
    - Skip the tests if neither of YAML or YAML::Syck is
      installed (but don't make it a test requirement,
      some people won't be using this with YAML at all).

0.01      2007-12-18
    - Initial release


t/lib/MXDefaultConfigTest.pm
package MXDefaultConfigTestBase;
use Moose;

has 'inherited_ro_attr' => (is => 'ro', isa => 'Str');

no Moose;
1;

package MXDefaultConfigTest;
use Moose;
extends 'MXDefaultConfigTestBase';
with 'MooseX::SimpleConfig';

has 'direct_attr' => (is => 'ro', isa => 'Int');

has 'req_attr' => (is => 'rw', isa => 'Str', required => 1);

has '+configfile' => ( default => 'test.yaml' );

no Moose;
1;


lib/MooseX/SimpleConfig.pm
package MooseX::SimpleConfig;
# ABSTRACT: A Moose role for setting attributes from a simple configuration file
# KEYWORDS: moose extension command line options attributes configuration file

our $VERSION = '0.12';

use Moose::Role;
with 'MooseX::ConfigFromFile';

use Config::Any 0.13 ();

sub get_config_from_file {
    my ($class, $file) = @_;

    $file = $file->() if ref $file eq 'CODE';
    my $files_ref = ref $file eq 'ARRAY' ? $file : [$file];

    my $can_config_any_args = $class->can('config_any_args');
    my $extra_args = $can_config_any_args ?
        $can_config_any_args->($class, $file) : {};
    ;
    my $raw_cfany = Config::Any->load_files({
        %$extra_args,
        use_ext         => 1,
        files           => $files_ref,
        flatten_to_hash => 1,
    } );

    my %raw_config;
    foreach my $file_tested ( reverse @{$files_ref} ) {
        if ( ! exists $raw_cfany->{$file_tested} ) {
            warn qq{Specified configfile '$file_tested' does not exist, } .
                qq{is empty, or is not readable\n};
                next;
        }

        my $cfany_hash = $raw_cfany->{$file_tested};
        die "configfile must represent a hash structure in file: $file_tested"
            unless $cfany_hash && ref $cfany_hash && ref $cfany_hash eq 'HASH';

        %raw_config = ( %raw_config, %{$cfany_hash} );
    }

    \%raw_config;
}

no Moose::Role; 1;

__END__

=pod

=head1 SYNOPSIS

  ## A YAML configfile named /etc/my_app.yaml:
  foo: bar
  baz: 123

  ## In your class
  package My::App;
  use Moose;

  with 'MooseX::SimpleConfig';

  has 'foo' => (is => 'ro', isa => 'Str', required => 1);
  has 'baz'  => (is => 'rw', isa => 'Int', required => 1);

  # ... rest of the class here

  ## in your script
  #!/usr/bin/perl

  use My::App;

  my $app = My::App->new_with_config(configfile => '/etc/my_app.yaml');
  # ... rest of the script here

  ####################
  ###### combined with MooseX::Getopt:

  ## In your class
  package My::App;
  use Moose;

  with 'MooseX::SimpleConfig';
  with 'MooseX::Getopt';

  has 'foo' => (is => 'ro', isa => 'Str', required => 1);
  has 'baz'  => (is => 'rw', isa => 'Int', required => 1);

  # ... rest of the class here

  ## in your script
  #!/usr/bin/perl

  use My::App;

  my $app = My::App->new_with_options();
  # ... rest of the script here

  ## on the command line
  % perl my_app_script.pl -configfile /etc/my_app.yaml -otherthing 123

=head1 DESCRIPTION

This role loads simple files to set object attributes.  It
is based on the abstract role L<MooseX::ConfigFromFile>, and uses
L<Config::Any> to load your configuration file.  L<Config::Any> will in
turn support any of a variety of different config formats, detected
by the file extension.  See L<Config::Any> for more details about
supported formats.

To pass additional arguments to L<Config::Any> you must provide a
C<config_any_args()> method, for example:

  sub config_any_args {
    return {
      driver_args => { General => { '-InterPolateVars' => 1 } }
    };
  }

Like all L<MooseX::ConfigFromFile> -derived file loaders, this
module is automatically supported by the L<MooseX::Getopt> role as
well, which allows specifying C<-configfile> on the command line.

=head1 ATTRIBUTES

=for stopwords configfile

=head2 configfile

Provided by the base role L<MooseX::ConfigFromFile>.  You can
provide a default configuration file pathname like so:

  has '+configfile' => ( default => '/etc/myapp.yaml' );

You can pass an array of filenames if you want, but as usual the array
has to be wrapped in a sub ref.

  has '+configfile' => ( default => sub { [ '/etc/myapp.yaml', '/etc/myapp_local.yml' ] } );

Config files are trivially merged at the top level, with the right-hand files taking precedence.

=head1 CLASS METHODS

=head2 new_with_config

Provided by the base role L<MooseX::ConfigFromFile>.  Acts just like
regular C<new()>, but also accepts an argument C<configfile> to specify
the file from which to load other attributes.  Explicit arguments
to C<new_with_config> will override anything loaded from the file.

=head2 get_config_from_file

Called internally by either C<new_with_config> or L<MooseX::Getopt>'s
C<new_with_options>.  Invokes L<Config::Any> to parse C<configfile>.

=cut
